NS linear solver Physics

GiudGiud opened the pull request:
wanted to see how hard it would be. Seems it wont be bad at all

left todo:
- [x] match all the tests for segregated-linear
- [x] add momentum gravity, only term with existing kernels

GiudGiud committed:
Add base physics for linear solver
refs #27280 #25642

GiudGiud committed:
Use segregated equations RC user object
Rename some routines
Use linear variables
Remove friction kernels code & friction attributes
refs #27280 #25642

GiudGiud committed:
Add pressure correction equation
Remove more of the nonlinear momentum terms
refs #27280 #25642

GiudGiud committed:
Add momentum linear discret. equation
Clean up, compiles
refs #27280 #25642

GiudGiud committed:
Add linear BC creation
refs #27280 #25642

GiudGiud committed:
Consolidate flow physics classes with a shared base class
refs #27280 #25642:wq

GiudGiud committed:
Add test for physics w/ linear solver
refs #27280 #25642

GiudGiud committed:
Add linear fv tasks to physics base
Add linear variable check to PhysicsBase class
refs #27280 #25642

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2352583/) on c5fc46c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28232/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28232/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9a7e0e1d7c085296a28c0506219b649d7779dd75`

GiudGiud committed:
Add test for physics w/ linear solver for a lid-driven cavity

GiudGiud committed:
Move some discretization attributes around between base flow and nonlinear flow Physics
Add two term BC for pressure to linear flow physics

GiudGiud committed:
Add support for gravity term in linear solver physics

GiudGiud committed:
Add doco for linear flow physics
refs #27280 #25642

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2353057/) on adf019d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28232/site)

This comment will be updated on new commits.

GiudGiud committed:
Polish rebase post k-eps physics merge and new wall boundary parameter

GiudGiud committed:
Improve physics documentation
- rework langugage
- add more hints to parameter checking
- add dump object problem trick

GiudGiud committed:
Fix up parameter checking for wall boundary parameter now being a map

GiudGiud committed:
Add support for momentum wall functors in linear flow physics
Mimick test setup more for comparing segregated linear fv inputs

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/ins/lid-driven/linear-segregated/lid-driven-segregated-physics.i:
```diff
@@ -0,0 +1,83 @@
+
+[Problem]
+  linear_sys_names = 'u_system v_system pressure_system'
+  previous_nl_solution_required = true
```
```suggestion
```

GiudGiud committed:
Small changes to enable TMAP8 PR
- give public access to list of variables for components
- const cast to be able to get non-const physics
- allow inheriting multi-species-diffusion

grmnptr commented on modules/navier_stokes/src/physics/WCNSFVLinearFlowPhysics.C:
```diff
@@ -0,0 +1,495 @@
+    const auto gravity_vector = getParam<RealVectorValue>("gravity");
+
+    for (const auto d : make_range(dimension()))
+    {
+      if (gravity_vector(d) != 0)
+      {
+        params.set<MooseFunctorName>("source_density") = std::to_string(gravity_vector(d));
+        params.set<NonlinearVariableName>("variable") = _velocity_names[d];
+
+        getProblem().addFVKernel(kernel_type, kernel_name + NS::directions[d], params);
+      }
+    }
```
```suggestion
      if (gravity_vector(d) != 0)
      {
        params.set<MooseFunctorName>("source_density") = std::to_string(gravity_vector(d));
        params.set<NonlinearVariableName>("variable") = _velocity_names[d];

        getProblem().addFVKernel(kernel_type, kernel_name + NS::directions[d], params);
      }
```

grmnptr commented on modules/navier_stokes/src/physics/WCNSFVLinearFlowPhysics.C:
```diff
@@ -0,0 +1,495 @@
+        getProblem().addLinearFVBC(bc_type, _velocity_names[d] + "_" + boundary_name, params);
+      }
+    }
+    else if (momentum_wall_type == "wallfunction")
+    {
+      // Placeholder
+      mooseError("Unsupported boundary condition type: " + std::string(momentum_wall_type));
+    }
+    else if (momentum_wall_type == "slip")
+    {
+      // Do nothing
+    }
+    else if (momentum_wall_type == "symmetry")
+    {
+      // Placeholder
+      mooseError("Unsupported boundary condition type: " + std::string(momentum_wall_type));
+    }
```
```suggestion
    else 
      mooseError("Unsupported wall boundary condition type: " + std::string(momentum_wall_type));
```

grmnptr replied:
Could be a param error too.

GiudGiud replied:
slip does not work?

grmnptr commented on modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/segregated-comparison/segregated-linear-physics.i:
```diff
@@ -0,0 +1,80 @@
+
+[Physics]
+  [NavierStokes]
+    [FlowLinear/flow]
```
Could be FlowSegregated(Linear). We will remove the old segregated capabilities once everything has been migrated to the new one. Or something similar.

This is just a conversation starter, the name could probably stay.

GiudGiud replied:
changed. FlowLinear was weird.

grmnptr commented on modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/segregated-comparison/segregated-linear-physics.i:
```diff
@@ -0,0 +1,80 @@
+[]
+
+[Problem]
+  #TODO: remove both
```
Yeah would be cool if we could do this under the hood. I think the previous solution flag is already in your code.

GiudGiud replied:
on it

GiudGiud replied:
actually this isnt easy. I dont think the problem should be created by the Physics, and this line creates the sytems. we would need to add APIs to make the problem create new systems from the Physics

grmnptr commented on modules/navier_stokes/test/tests/finite_volume/ins/lid-driven/linear-segregated/lid-driven-segregated-physics.i:
```diff
@@ -0,0 +1,82 @@
+[]
+
+
+[Physics]
```
Nice syntax!

grmnptr reviewed (changes requested):
Pretty neat!

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2470132/), step Docs: sync website on e72135d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28232/site)

This comment will be updated on new commits.

GiudGiud commented on modules/navier_stokes/src/physics/WCNSFVFlowPhysicsBase.C:
```diff
@@ -0,0 +1,418 @@
+  params.transferParam<MooseEnum>(NSFVBase::validParams(), "velocity_interpolation");
+  params.transferParam<MooseEnum>(NSFVBase::validParams(), "momentum_advection_interpolation");
+  params.transferParam<bool>(NSFVBase::validParams(), "pressure_two_term_bc_expansion");
+  params.transferParam<Real>(INSFVMomentumAdvection::validParams(), "characteristic_speed");
```
ok I ll suppress

GiudGiud replied:
or move the transfer

GiudGiud commented on modules/navier_stokes/src/physics/WCNSFVFlowPhysicsBase.C:
```diff
@@ -0,0 +1,418 @@
+unsigned short
+WCNSFVFlowPhysicsBase::getNumberAlgebraicGhostingLayersNeeded() const
+{
+  unsigned short ghost_layers = 2;
```
that's great. I ll override

GiudGiud commented on modules/navier_stokes/doc/content/source/physics/WCNSFVLinearFlowPhysics.md:
```diff
@@ -0,0 +1,65 @@
+
+## Automatically defined variables
+
+The `WCNSFVLinearFlowPhysics` automatically sets up the variables which are
```
good point. will do

GiudGiud replied:
done

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/segregated-comparison/segregated-linear-physics.i:
```diff
@@ -0,0 +1,80 @@
+[Problem]
+  #TODO: remove both
+  linear_sys_names = 'u_system v_system pressure_system'
+  previous_nl_solution_required = true
```
```suggestion
```

GiudGiud committed:
Address Peter's review
- rename FVLinear to LinearFV
- use FlowSegregated
- more accurate doco
- remove more parameters not handled by segregated physics yet

Co-authored-by: Peter German <31662443+grmnptr@users.noreply.github.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2472152/), step Clang format on df4778e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28232/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28232/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a31ebecb4216bdedfbd1d6ad3ec6c94969c03c07`

GiudGiud committed:
Remove symlinks for syntax, bad practice
Update flowLinear to FlowSegregated

GiudGiud commented on modules/navier_stokes/include/physics/WCNSFVLinearFlowPhysics.h:
```diff
@@ -0,0 +1,66 @@
+
+  UserObjectName rhieChowUOName() const override;
+
+  /// Name of the vector to hold pressure momentum equation contributions
```
removed

GiudGiud committed:
Address Peter s review
- 1 ghost layer for linear FV
- rename from NonLinear/linear to solver variables

Fix adding old/older solutions from the physics

GiudGiud commented on framework/src/physics/PhysicsBase.C:
```diff
@@ -276,6 +276,21 @@ PhysicsBase::copyVariablesFromMesh(const std::vector<VariableName> & variables_t
+               name(),
+               "' but it is already defined as auxiliary");
+  else
+    return false;
```
changes to variableExists

at this point this is more what we want to know, but later on we might want to add a boolean for linear or not

GiudGiud committed:
Fixes for turbulence test

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2473668/), step Generate coverage on 37ec70a wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>bfd459</th>
      <th colspan=4>#28232 37ec70</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.05%</td>
      <td>85.05%</td>
      <td>+0.00%</td>
      <td>90.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>106006</td>
      <td>106007</td>
      <td>+1</td>
      <td>18</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18639</td>
      <td>18639</td>
      <td>-</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28232/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28232/coverage/framework)


## Modules coverage

### Heat transfer

<table>
  <thead>
    <tr>
      <th></th>
      <th>bfd459</th>
      <th colspan=4>#28232 37ec70</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.01%</td>
      <td>89.01%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4275</td>
      <td>4275</td>
      <td>-</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>528</td>
      <td>528</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28232/coverage_diff/heat_transfer)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28232/coverage/heat_transfer)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>bfd459</th>
      <th colspan=4>#28232 37ec70</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.59%</td>
      <td>84.48%</td>
      <td>-0.11%</td>
      <td>85.53%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>16765</td>
      <td>16997</td>
      <td>+232</td>
      <td>473</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3055</td>
      <td>3123</td>
      <td>+68</td>
      <td>80</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28232/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28232/coverage/navier_stokes)

### Scalar transport

<table>
  <thead>
    <tr>
      <th></th>
      <th>bfd459</th>
      <th colspan=4>#28232 37ec70</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.44%</td>
      <td>88.44%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>352</td>
      <td>352</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>46</td>
      <td>46</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28232/coverage_diff/scalar_transport)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28232/coverage/scalar_transport)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28232/coverage/xfem)

</details>

## Warnings

- `navier_stokes` new line coverage rate 85.53% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud committed:
Documentation fixes

grmnptr commented on modules/navier_stokes/include/physics/WCNSLinearFVFlowPhysics.h:
```diff
@@ -0,0 +1,66 @@
+#include "WCNSFVTurbulencePhysics.h"
+
+/**
+ * Creates all the objects needed to solve the Navier-Stokes with the SIMPLE algorithm
```
```suggestion
 * Creates all the objects needed to solve the Navier-Stokes equations with the SIMPLE algorithm
```

grmnptr commented on modules/navier_stokes/src/physics/WCNSLinearFVFlowPhysics.C:
```diff
@@ -0,0 +1,492 @@
+    if (momentum_outlet_type == "zero-gradient" || momentum_outlet_type == "fixed-pressure" ||
+        momentum_outlet_type == "fixed-pressure-zero-gradient")
+    {
+      {
+        const std::string bc_type = "LinearFVAdvectionDiffusionOutflowBC";
+        InputParameters params = getFactory().getValidParams(bc_type);
+        params.set<std::vector<BoundaryName>>("boundary") = {outlet_bdy};
+
+        for (const auto d : make_range(dimension()))
+        {
+          params.set<LinearVariableName>("variable") = _velocity_names[d];
+          getProblem().addLinearFVBC(bc_type, _velocity_names[d] + "_" + outlet_bdy, params);
+        }
+      }
```
```suggestion
        const std::string bc_type = "LinearFVAdvectionDiffusionOutflowBC";
        InputParameters params = getFactory().getValidParams(bc_type);
        params.set<std::vector<BoundaryName>>("boundary") = {outlet_bdy};

        for (const auto d : make_range(dimension()))
        {
          params.set<LinearVariableName>("variable") = _velocity_names[d];
          getProblem().addLinearFVBC(bc_type, _velocity_names[d] + "_" + outlet_bdy, params);
        }
```

grmnptr reviewed (changes requested):
I was not able to add comments to the file below so I'll just list them here:
- FluidHeatTransferPhyis -> link in text and syntax links might not be correct
- ScalarTransportPhys -> syntax links might not be correct
- SolidHeatTransferPhys -> syntax links might not be correct
- Trubulence -> syntax links might not be correct
- TwoPhaseMixturePhys -> syntax links might not be correct
Looks nice otherwise!

GiudGiud committed:
Address Peter's review:
- remove unnecessary bracketing
- fix syntax listings for NS

Co-authored-by: Peter German <31662443+grmnptr@users.noreply.github.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2475580/), step Clang format on db19e92 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28232/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28232/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format bfd4591af3f9696861c382c1a33257b975e5c9ef`

GiudGiud commented:
good catch I had never looked at these syntax listings

GiudGiud commented:
Test failures unrelated imo

grmnptr reviewed (approved):
Nice!

GiudGiud commented:
Thanks for the review!
